langchain/src/vectorstores/prisma.ts

```diff
@@ -0,0 +1,249 @@
+};
+
+type ModelColumns<TModel extends Record<string, unknown>> = {
+  [K in keyof TModel]?: true | ColumnSymbol;
+};
+
+type SimilarityModel<
+  TModel extends Record<string, unknown> = Record<string, unknown>,
+  TColumns extends ModelColumns<TModel> = ModelColumns<TModel>
+> = Pick<TModel, keyof ObjectIntersect<TModel, TColumns>> & {
+  _distance: number | null;
+};
+
+export class PrismaVectorStore<
+  TModel extends Record<string, unknown>,
+  TModelName extends string,
+  TSelectModel extends ModelColumns<TModel>
+> extends VectorStore {
+  tableSql: Sql;
+
+  vectorColumnSql: Sql;
+
+  selectSql: Sql;
+
+  idColumn: keyof TModel & string;
+
+  contentColumn: keyof TModel & string;
+
+  static IdColumn: typeof IdColumnSymbol = IdColumnSymbol;
+
+  static ContentColumn: typeof ContentColumnSymbol = ContentColumnSymbol;
+
+  protected db: PrismaClient;
+
+  protected Prisma: PrismaNamespace;
+
+  constructor(
+    config: {
+      db: PrismaClient;
+      prisma: PrismaNamespace;
+      tableName: TModelName;
+      vectorColumnName: string;
+      columns: TSelectModel;
+    },
+    embeddings: Embeddings
+  ) {
+    super(embeddings, {});
+
+    this.Prisma = config.prisma;
+    this.db = config.db;
+
+    const entries = Object.entries(config.columns);
+    const idColumn = entries.find((i) => i[1] === IdColumnSymbol)?.[0];
+    const contentColumn = entries.find(
+      (i) => i[1] === ContentColumnSymbol
+    )?.[0];
+
+    if (idColumn == null) throw new Error("Missing ID column");
+    if (contentColumn == null) throw new Error("Missing content column");
+
+    this.idColumn = idColumn;
+    this.contentColumn = contentColumn;
+
+    this.tableSql = this.Prisma.raw(`"${config.tableName}"`);
+    this.vectorColumnSql = this.Prisma.raw(`"${config.vectorColumnName}"`);
+
+    this.selectSql = this.Prisma.raw(
+      entries
+        .map(([key, alias]) => (alias && key) || null)
+        .filter((x): x is string => !!x)
+        .map((key) => `"${key}"`)
+        .join(", ")
+    );
+  }
+
+  static withModel<TModel extends Record<string, unknown>>(db: PrismaClient) {
+    function create<
+      TPrisma extends PrismaNamespace,
+      TColumns extends ModelColumns<TModel>
+    >(
+      config: {
+        prisma: TPrisma;
+        tableName: keyof TPrisma["ModelName"] & string;
+        vectorColumnName: string;
+        columns: TColumns;
+      },
+      embeddings: Embeddings
+    ) {
+      type ModelName = keyof TPrisma["ModelName"] & string;
+      return new PrismaVectorStore<TModel, ModelName, TColumns>(
+        { db, ...config },
+        embeddings
+      );
+    }
+
+    return { create };
+  }
+
+  async addModels(models: TModel[]) {
+    return this.addDocuments(
+      models.map((metadata) => {
+        const pageContent = typeof metadata[this.contentColumn];
+        if (pageContent !== "string")
+          throw new Error("Content column must be a string");
+        return new Document({ pageContent, metadata });
```

Comment:
I think we'd want to remove the pageContent from `metadata`?

Reply:
At the moment, it returns the subset a Prisma model (`TModel` and `TSelectModel`), which do include the `this.contentColumn`). 

We could remove the content column from the metadata, but I would argue that the metadata should contain the same columns, as user specifies in the `dbConfig`.